src/sentry/ingest/billing_metrics_consumer.py

```diff
@@ -108,35 +121,174 @@ def __init__(
+        self._mark_commit_ready()
+        self._bulk_commit()
+
+        while self._ongoing_billing_outcomes:
+            now = time.time()
+            time_left = deadline - now if deadline else None
+
+            if time_left is not None and time_left <= 0:
+                items_left = self._messages_ready_since_last_commit + len(
+                    self._ongoing_billing_outcomes
+                )
+                logger.warning(f"join timed out, items left in the queue: {items_left}")
+                break
+
+            future, metrics_msg = self._ongoing_billing_outcomes[0]
+            do_commit = True
+            if future:
+                try:
+                    future.result(time_left)
+                except Exception:
+                    logger.error(
+                        "Async future failed in billing metrics consumer.",
+                        exc_info=future.exception(),
+                        extra={"offset": metrics_msg.offset},
+                    )
+                    # Don't commit the future when it fails -- if no futher messages
+                    # are committed before shutting down the consumer, new consumers
+                    # will process the current offset again.
+                    do_commit = False
+                    self._ongoing_billing_outcomes.popleft()
+
+            if do_commit:
+                self._mark_commit_ready()
+                self._bulk_commit()
+
+        self._billing_producer.close()
+
     def terminate(self) -> None:
         self.close()
+        amount_dropped = len(self._ongoing_billing_outcomes)
+        while self._ongoing_billing_outcomes:
+            ongoing_work, _ = self._ongoing_billing_outcomes.popleft()
+            if ongoing_work:
+                ongoing_work.cancel()
+        amount_dropped += self._clear_ready_queue()
+        if amount_dropped > 0:
+            logger.warning(f"terminated, items dropped: {amount_dropped}")
+
+        self._billing_producer.close()
 
     def close(self) -> None:
-        self.__closed = True
+        self._closed = True
 
     def submit(self, message: Message[KafkaPayload]) -> None:
-        assert not self.__closed
-        self.__messages_since_last_commit += 1
+        assert not self._closed
 
-        payload = self._get_payload(message)
-        self._produce_billing_outcomes(payload)
-        self._mark_commit_ready(message)
+        bucket_payload = self._get_bucket_payload(message)
+        quantity = self._count_processed_transactions(bucket_payload)
 
-    def _get_payload(self, message: Message[KafkaPayload]) -> MetricsBucket:
+        if quantity < 1:
+            # We still want to commmit buckets that don't generate billing
+            # outcomes, since the consumer has already processed them. To keep
+            # the offset order when committing, we still need to go through the
+            # futures queue.
+            future = None
+        else:
+            value = json.dumps(
```

Comment:
Since you use rapidjson to load it, i would also use rapidjson here

Reply:
Apparently, there's no rapidjson in dumps